### PR TITLE
Debug/decode llama8 b perf dp4 rebased stable

### PR DESCRIPTION
**Pull request by mbahnasTT**
### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:{{branch_name}})
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:{{branch_name}})
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:{{branch_name}})
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs:
      [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and
      [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
      and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs:
      [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset
      (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and
      [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs:
      [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset
      (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and
      [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: models/common/sampling/generator.py [hunk 1]
```diff
@@ -108,20 +103,26 @@ def reset_output_state(self, tokens=None):
...
             k=sampling_params.top_k,
             p=sampling_params.top_p,
             temp=sampling_params.temperature,
             enable_log_probs=sampling_params.enable_log_probs,
         )
-        self.tt_penalties.reset_params(
-            sampling_params.presence_penalty, sampling_params.frequency_penalty, sampling_params.repetition_penalty
-        )
+        if self.tt_sampling._force_argmax_sampling != old_force_argmax_sampling:
+            self.reset_trace()
```
Comment (bgoelTT) [inline models/common/sampling/generator.py]: It doesn't look like `self.reset_trace()` is being called here because `old_force_argmax_sampling` is initialized to `self.tt_sampling._force_argmax_sampling`.

File: models/common/sampling/generator.py [hunk 2]
```diff
@@ -227,38 +230,39 @@ def sample(
...
-        return tt_out
+        return tt_out, tt_log_probs
 
     def reset_seed(self, seed):
         for i, s in enumerate(seed):
             if s is None:
-                # set to random seed to have variability while using tensor manual_seed
-                seed[i] = random.randint(0, 1000000)
+                # set to default seed value which is 0
+                seed[i] = 0
```
Comment (bgoelTT) [inline models/common/sampling/generator.py]: This change now means that random sampling has fixed seed by default, compared to the previous, and expected behaviour that the default seeding is random unless specified in the input SamplingParams

Comment (sraizada-tt) [reply models/common/sampling/generator.py]: I don't think we want this because this will diverge from vllm's compat sampling which says seed=None means non-determinism

**Comment by yieldthought** [comment]
Codex review:

[P2] Preserve decode output return type for non-logprob callers - [models/demos/llama3_70b_galaxy/tt/generator.py:751](https://github.com/tenstorrent/tt-metal/blob/ed89ecdc26c299dc04ea46245b6fda66e019ac3f/models/demos/llama3_70b_galaxy/tt/generator.py#L751)
The updated `process_decode_output_host` now always returns `(tokens_or_logits, log_probs)` even when log-probs aren't requested. Callers like `[models/demos/llama3_70b_galaxy/demo/text_qwen_demo.py:833](https://github.com/tenstorrent/tt-metal/blob/ed89ecdc26c299dc04ea46245b6fda66e019ac3f/models/demos/llama3_70b_galaxy/demo/text_qwen_demo.py#L833)` and the subsequent `out_tok.shape` usage at `[models/demos/llama3_70b_galaxy/demo/text_qwen_demo.py:837](https://github.com/tenstorrent/tt-metal/blob/ed89ecdc26c299dc04ea46245b6fda66e019ac3f/models/demos/llama3_70b_galaxy/demo/text_qwen_demo.py#L837)` still treat the return value as a tensor, so running the Qwen decode demo with async read will now throw an attribute error when it receives a tuple instead of a tensor. This needs either a compatibility branch (return a tensor when log-probs are disabled) or corresponding updates to the remaining call sites.

**Review by yieldthought — APPROVED** [review]
Approved to unblock but please take a look at the codex comment regarding return types
